### PR TITLE
feat(model): add Muse Glimmer support

### PR DESCRIPTION
**Pull request by yaoyu-33**
# What does this PR do?

Adds Muse Glimmer 30B support through the builder-backed model config, model builder, and model bridge APIs.

## Modeling

`MuseGlimmerModel` directly subclasses Megatron-Core `HybridModel`. `MuseGlimmerModelBuilder` constructs it with a model-local `HybridStack` specification whose `*` layer pattern selects the complete Muse transformer layer. The implementation does not build, wrap, alias, or present the text model as `GPTModel`.

The text path preserves Muse's 3:1 sliding/full attention pattern, per-layer NoPE pattern, GQA, weightless Q/K RMSNorm, full-head sigmoid attention gate, centered pre/post residual norms, final scaled RMSNorm, normalized embeddings, output multiplier, and tanh logit soft-cap. Muse-specific behavior remains confined to the Muse Glimmer model family while the stack uses Megatron-Core transformer components.

The replicated vision path implements the 50-layer encoder, learned-position lookup, 2-D RoPE, window/full attention, pixel shuffle, adapter, projection, and perception RMSNorm. Its parameters are marked for tensor-parallel gradient averaging.

The bridge maps the complete checkpoint against native HybridModel parameter roots, including model-local Q/K/V/full-head gate fusion and split semantics.

# Changelog

- Add `MuseGlimmerModelConfig`, `MuseGlimmerModelBuilder`, and the combined VLM.
- Build the language stack as a native Megatron-Core `HybridModel`/`HybridStack`.
- Add complete HF ↔ Megatron mappings and AutoBridge registration.
- Add CPU and builder-backed checkpoint conversion support.
- Add deterministic builder-backed VLM inference support.
- Use the official Transformers 5.15.0 Muse Glimmer configs directly and keep the native HybridModel config in `muse_glimmer_config.py`.
- Add focused config, mapping, checkpoint, norm, vision, HybridModel construction, and registration tests.
- Add modeling documentation, training recipes, and a model verification card.

# Validation

Final DCO-signed head: `a80ab5709f1f04982c4ce907729d6b524c0a3f2b`.

- CPU and TP8 HF → Megatron imports each completed all 1,228 mappings and persisted/reloaded a native 52-layer `HybridModel` with 29,776,626,688 parameters.
- CPU and TP8 Megatron → HF exports each passed an exact audit over all 1,436 BF16 tensors and 29,776,626,688 elements: zero missing, unexpected, shape, dtype, or value mismatches and maximum absolute difference 0. Transformers 5.15.0 strictly reloaded both exports as `MuseGlimmerForConditionalGeneration` with the same parameter count.
- TP8 HF/Megatron forward comparison predicted the same next token. Logit cosine similarity was 0.999896; maximum and mean absolute differences were 0.781250 and 0.139148.
- Two independent deterministic TP8 greedy runs produced the same exact 64-token image-conditioned completion.
- Bounded 32-H100 multimodal pretraining completed 100 finite optimizer steps with zero skipped or NaN iterations, loss `12.23977 -> 1.4103`, and complete step-50/100 checkpoints.
- Full-model SFT, post-SFT export/inference, and packed 8K CP2 SFT completed their bounded verification runs with finite loss and complete reloadable artifacts.
- TP8 LoRA completed 100 finite optimizer steps with zero skipped or NaN iterations, saved all 208 expected adapters, and passed a direct step-100 reload followed by finite step 101 and a distinct complete save.
- The canonical 32-H100 FP8 recipe completed 50 finite steps and averaged 428.24 model TFLOP/s/GPU over its final 10 steps, satisfying the requested 400-500 range.
- Direct pretraining resume restored full state and saved a complete distinct 32-shard checkpoint, but the final loss sentinel differed by 2.9638%, exceeding the fixed 1% gate. The card therefore leaves checkpoint resume explicitly unverified; the tolerance was not weakened.
- Original focused Muse HybridModel, conversion mapping, checkpoint schema, deterministic inference, and related common-path validation: 85 passed. Post-cleanup/current-main focused validation: 116 passed.
- Final current-main focused regression passed 499 tests with 1 skip across the original DeepSeek failure, Muse HybridModel/config/mappings/recipes, Gemma Transformers 5.15 paths, builder conversion/inference, checkpointing, shared parameter mappings, direct-HF SFT, docs consistency, and verification-card validation.
- Model verification card validator: passed.
- `uv run pre-commit run --all-files`: passed.
- Independent final review found no blocking or high-severity issues. Its embedding-sharing config finding was fixed by honoring the top-level VLM contract and adding a divergent nested/top-level regression test.
- Current upstream `main` at `74749b4f2e456800255d65d2735c522443f9c384` was inspected and is contained in the final head.

The full conversion, forward, inference, PEFT, and performance evidence is recorded in `examples/model_verification_cards/muse-glimmer-30b/card.yaml`.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Read and followed the contributor guidelines.
- [x] Added focused tests.
- [x] Added modeling documentation and a model verification card.
- [x] No new dependency was added; the existing Transformers requirement was updated to 5.15.0 for native Muse Glimmer support.

# Additional Information

- Public model: https://huggingface.co/meta-models/Muse-Glimmer-30B
- Public architecture overview: https://huggingface.co/blog/muse-glimmer

### CODE REVIEW

File: examples/conversion/compare_hf_and_megatron/compare.py [hunk 1]
```diff
@@ -694,14 +694,26 @@ def _load_megatron_model(args):
...
         )
-        model_provider = bridge.to_megatron_provider(load_weights=False)
-        model_provider.tensor_model_parallel_size = tp
-        model_provider.pipeline_model_parallel_size = pp
-        model_provider.expert_model_parallel_size = ep
-        model_provider.expert_tensor_parallel_size = etp
-        model_provider.pipeline_dtype = torch.bfloat16
-        model_provider.finalize()
-        model_provider.initialize_model_parallel(seed=0)
+        if getattr(bridge._model_bridge, "MODEL_CONFIG_CLASS", None) is not None:
```
Comment (cuichenx) [inline examples/conversion/compare_hf_and_megatron/compare.py]: why is this needed for muse glimmer?

File: src/megatron/bridge/models/muse_glimmer/modeling_muse_glimmer.py [hunk 2]
```diff
@@ -0,0 +1,795 @@
...
+    slice_batch_for_context_parallel,
+)
+
+
+if TYPE_CHECKING:
+    from megatron.core.inference.contexts import BaseInferenceContext
+    from megatron.core.packed_seq_params import PackedSeqParams
+
+
+def _rms_norm(hidden_states: Tensor, eps: float) -> Tensor:
```
Comment (cuichenx) [inline src/megatron/bridge/models/muse_glimmer/modeling_muse_glimmer.py]: do we want to do this for parity instead of using TE norm? this would make perf really bad

File: examples/conversion/compare_hf_and_megatron/compare.py (hunk 1 above)
Comment (yaoyu-33) [reply examples/conversion/compare_hf_and_megatron/compare.py]: Muse Glimmer is the first model in this comparison path that is builder-backed (`MODEL_CONFIG_CLASS`) instead of provider-backed. The branch is capability-based, not Muse-specific: it builds the outer model config, applies runtime parallelism to its nested `transformer`, and initializes process groups through the builder contract. The TP8 model-comparison card run exercises this path.

File: src/megatron/bridge/models/muse_glimmer/modeling_muse_glimmer.py (hunk 2 above)
Comment (yaoyu-33) [reply src/megatron/bridge/models/muse_glimmer/modeling_muse_glimmer.py]: Addressed in `c64a34dcb` and `923f894c0`: every affine decoder RMSNorm (including zero-centered gamma and final norm) and every affine vision LayerNorm now uses `TENorm` whenever TE is available, independently of the attention/linear backend. Only the architecture-defined weightless Q/K, embedding, and perception RMS norms stay parameter-free in torch because the pinned TE RMSNorm always owns a learnable gamma; using it there would change checkpoint and optimizer schemas. The TP8 forward rerun passes with matching next token and 0.999911 cosine.
